File: Data Structures & Algorithms/binary-search/submission-3.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int l = 0;
        int r = nums.size() - 1;
        while (l <= r) {
            int k = (l + r) / 2;

            
            if (nums[k] > target) {
                r = k - 1;
            } 
            else if(nums[k] < target) {
                l = k + 1;
            }
            else if(nums[k] == target)
            {
                return k;
            }
        }
        return -1;
    }
};

File: Data Structures & Algorithms/sliding-window-maximum/submission-3.py
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        dic = {}
        current = 0
        answer = []

        if len(nums) < k:
            return [max(nums)]

        for i in range(k ):
            dic[nums[i]] = dic.get(nums[i], 0) + 1
        
        # print(f"sum dic val is {sum(dic.values())}")

        while current < len(nums) - k + 1:
            # print("=============new==============")
            # for key, value in dic.items():
                # print(f"key: {key}, value {value}")

            answer.append(max(dic))
            dic[nums[current]] = dic[nums[current]] - 1
            if dic[nums[current]] == 0:
                del dic[nums[current]]
            
            if current + k <= len(nums) - 1:
                # print(f"insterting {nums[current + k]}")
                dic[nums[current + k]] = dic.get(nums[current + k], 0) + 1

            current += 1
            






        return answer

        

